File: CPU/GeneralUnion/HyperLogLog.java
import java.util.BitSet;
import java.util.HashSet;
import java.util.Random;

public class HyperLogLog extends GeneralDataStructure {
	/** parameters for HLL */
	public String name = "HyperLogLog";				// the data structure name
	public int HLLSize; 								// unit size for each register
	public int m;									// the number of registers in the HLL sketch
	public int maxRegisterValue;
	public double alpha;
	public BitSet[] HLLMatrix;
	
	public HyperLogLog(int m, int size) {
		super();
		HLLSize = size;
		this.m = m;
		maxRegisterValue = (int) (Math.pow(2, size) - 1);
		HLLMatrix = new BitSet[m];
		for (int i = 0; i < m; i++) {
			HLLMatrix[i] = new BitSet(size);
		}
		alpha = getAlpha(m);
	}

	@Override
	public String getDataStructureName() {
		return name;
	}
	
	@Override
	public BitSet getBitmaps() {
		return null;
	}
	
	@Override
	public int[] getCounters() {
		return null;
	};
	
	@Override
	public BitSet[] getFMSketches() {
		return null;
	};
	
	@Override
	public BitSet[] getHLLs() {
		return HLLMatrix;
	};

	@Override
	public int getUnitSize() {
		return HLLSize*m;
	}

	@Override
	public int getValue() {
		return getValue(HLLMatrix);
	}
	@Override
	public GeneralDataStructure[] getsplit(int m,int w) {
		GeneralDataStructure[]  B=new HyperLogLog[m];
		for(int i=0;i<m;i++) {
			B[i]=new HyperLogLog(w,5);
			for(int j=0;j<w/m;j++)
				B[i].getHLLs()[j]=this.HLLMatrix[i*(w/m)+j];
		}
		return B;
	}
	@Override
	public void encode() {
		int r = rand.nextInt(Integer.MAX_VALUE);
		int k = r % m;
		int hash_val = GeneralUtil.intHash(r);
		int leadingZeros = Integer.numberOfLeadingZeros(hash_val) + 1; 
		leadingZeros = Math.min(leadingZeros, maxRegisterValue);
		if (getBitSetValue(HLLMatrix[k]) < leadingZeros) {
			setBitSetValue(k, leadingZeros);
		}
	}
	
	@Override
	public void encode(String elementID) {
		int hash_val = GeneralUtil.FNVHash1(elementID);
		int leadingZeros = Integer.numberOfLeadingZeros(hash_val) + 1;
		int k = (hash_val % m + m) % m;
		leadingZeros = Math.min(leadingZeros, maxRegisterValue);
		if (getBitSetValue(HLLMatrix[k]) < leadingZeros) {
			setBitSetValue(k, leadingZeros);
		}
	}
	
	@Override
	public void encode(int elementID) {
 		int hash_val = GeneralUtil.intHash(elementID);
		int leadingZeros = Integer.numberOfLeadingZeros(hash_val) + 1;
		int k = (hash_val % m + m) % m;
		leadingZeros = Math.min(leadingZeros, maxRegisterValue);
		if (getBitSetValue(HLLMatrix[k]) < leadingZeros) {
			setBitSetValue(k, leadingZeros);
		}
	}
	
	
	@Override
	public int getValue(String flowID, int[] s) {
		int ms = s.length;
		BitSet[] sketch = new BitSet[ms];
		for (int j = 0; j < ms; j++) {
			int i = (GeneralUtil.intHash(GeneralUtil.FNVHash1(flowID) ^ s[j]) % m + m) % m;
			sketch[j] = HLLMatrix[i];
		}
		return getValue(sketch);
	}
	
	public int getValue(long flowID, int[] s) {
		int ms = s.length;
		BitSet[] sketch = new BitSet[ms];
		for (int j = 0; j < ms; j++) {
			int i = (GeneralUtil.intHash(GeneralUtil.FNVHash1(flowID) ^ s[j]) % m + m) % m;
			sketch[j] = HLLMatrix[i];
		}
		return getValue(sketch);
	}
	
	public int getValue(BitSet[] sketch) {
		Double result = 0.0;
		int zeros = 0;
		int len = sketch.length;
		for (int i = 0; i < len; i++) {
			if (getBitSetValue(sketch[i]) == 0) zeros++;
			result = result + Math.pow(2, -1.0 * getBitSetValue(sketch[i]));
		}
		result = alpha * len * len / result;
		if (result <= 5.0 / 2.0 * len) {			// small flows
			result = 1.0 * len * Math.log(1.0 * len / Math.max(zeros, 1));
		} else if (result > Integer.MAX_VALUE / 30.0) {
			result = -1.0 * Integer.MAX_VALUE * Math.log(1 - result / Integer.MAX_VALUE);
		}
		return result.intValue();
	}
	
	public static int getBitSetValue(BitSet b) {
		int res = 0;
		for (int i = 0; i < b.length(); i++) {
			if (b.get(i)) res += Math.pow(2, i);
		}
		return res;
	}
	
	public void setBitSetValue(int index, int value) {
		int i = 0;
		while (value != 0 && i < HLLSize) {
			if ((value & 1) != 0) {
				HLLMatrix[index].set(i);
			} else {
				HLLMatrix[index].clear(i);
			}
			value = value >>> 1;
			i++;
		}
	}
	
	public double getAlpha(int m) {
		double a;
		if (m == 16) {
			a = 0.673;
		} else if (m == 32) {
			a = 0.697;
		} else if (m == 64) {
			a = 0.709;
		} else {
			a = 0.7213 / (1 + 1.079 / m);
		}
		return a;
	}
	@Override
	public void encodeSegment(long flowID, long elementID, int[] s, int w) {
		int ms = s.length;	
		int j = ((GeneralUtil.FNVHash1(flowID^elementID)) % ms + ms) % ms;
		int k = (GeneralUtil.intHash(GeneralUtil.FNVHash1(flowID) ^ s[j]) % w + w) % w;
		int i = j * w + k;
		int hash_val = GeneralUtil.FNVHash1(elementID);
		int leadingZeros = Integer.numberOfLeadingZeros(hash_val) + 1;
		leadingZeros = Math.min(leadingZeros, maxRegisterValue);
		if (getBitSetValue(HLLMatrix[i]) < leadingZeros) {
			setBitSetValue(i, leadingZeros);
		}
		
	}
	@Override
	public void encodeSegment(String flowID, String elementID, int[] s, int w) {
		int ms = s.length;
		int j = ((GeneralUtil.FNVHash1(elementID)^GeneralUtil.FNVHash1(flowID)) % ms + ms) % ms;
		int k = (GeneralUtil.intHash(GeneralUtil.FNVHash1(flowID) ^ s[j]) % w + w) % w;
		int i = j * w + k;
		int hash_val = GeneralUtil.FNVHash1(elementID);
		int leadingZeros = Integer.numberOfLeadingZeros(hash_val) + 1;
		leadingZeros = Math.min(leadingZeros, maxRegisterValue);
		if (getBitSetValue(HLLMatrix[i]) < leadingZeros) {
			setBitSetValue(i, leadingZeros);
		}
	}

	@Override
	public void encodeSegment(int flowID, int elementID, int[] s, int w) {
		int ms = s.length;
		int j = (GeneralUtil.intHash(elementID) % ms + ms) % ms;
		int k = (GeneralUtil.intHash(flowID ^ s[j]) % w + w) % w;
		int i = j * w + k;
		int hash_val = GeneralUtil.intHash(elementID);
		int leadingZeros = Integer.numberOfLeadingZeros(hash_val) + 1;
		leadingZeros = Math.min(leadingZeros, maxRegisterValue);
		if (getBitSetValue(HLLMatrix[i]) < leadingZeros) {
			setBitSetValue(i, leadingZeros);
		}
	}

	@Override
	public void encodeSegment(String flowID, int[] s, int w) {
		int ms = s.length;
		int r = rand.nextInt(Integer.MAX_VALUE);
		int j = r % ms;
		int k = (GeneralUtil.intHash(GeneralUtil.FNVHash1(flowID) ^ s[j]) % w + w) % w;
		int i = j * w + k;
		int hash_val = GeneralUtil.FNVHash1(String.valueOf(r));
		int leadingZeros = Integer.numberOfLeadingZeros(hash_val) + 1;
		leadingZeros = Math.min(leadingZeros, maxRegisterValue);
		if (getBitSetValue(HLLMatrix[i]) < leadingZeros) {
			setBitSetValue(i, leadingZeros);
		}
	}
	
	@Override
	public void encodeSegment(long flowID, int[] s, int w) {
		int ms = s.length;
		int r = rand.nextInt(Integer.MAX_VALUE);
		int j = r % ms;
		int k = (GeneralUtil.intHash(GeneralUtil.FNVHash1(flowID) ^ s[j]) % w + w) % w;
		int i = j * w + k;
		int hash_val = GeneralUtil.FNVHash1(String.valueOf(r));
		int leadingZeros = Integer.numberOfLeadingZeros(hash_val) + 1;
		leadingZeros = Math.min(leadingZeros, maxRegisterValue);
		if (getBitSetValue(HLLMatrix[i]) < leadingZeros) {
			setBitSetValue(i, leadingZeros);
		}
	}

	@Override
	public void encodeSegment(int flowID, int[] s, int w) {
		int ms = s.length;
		int r = rand.nextInt(Integer.MAX_VALUE);
		int j = r % ms;
		int k = (GeneralUtil.intHash(flowID ^ s[j]) % w + w) % w;
		int i = j * w + k;
		int hash_val = GeneralUtil.intHash(r);
		int leadingZeros = Integer.numberOfLeadingZeros(hash_val) + 1;
		leadingZeros = Math.min(leadingZeros, maxRegisterValue);
		if (getBitSetValue(HLLMatrix[i]) < leadingZeros) {
			setBitSetValue(i, leadingZeros);
		}
	}
	
	@Override
	public int getValueSegment(String flowID, int[] s, int w) {
		int ms = s.length;
		BitSet[] sketch = new BitSet[ms];
		for (int j = 0; j < ms; j++) {
			int i = j * w + (GeneralUtil.intHash(GeneralUtil.FNVHash1(flowID) ^ s[j]) % w + w) % w;
			sketch[j] = HLLMatrix[i];
		}
		int result = getValue(sketch);
		return result;
	}
	@Override
	public int getValueSegment(long flowID, int[] s, int w) {
		int ms = s.length;
		BitSet[] sketch = new BitSet[ms];
		for (int j = 0; j < ms; j++) {
			int i = j * w + (GeneralUtil.intHash(GeneralUtil.FNVHash1(flowID) ^ s[j]) % w + w) % w;
			sketch[j] = HLLMatrix[i];
		}
		int result = getValue(sketch);
		return result;
	}
	
	
	@Override
	public GeneralDataStructure join(GeneralDataStructure gds) {
		HyperLogLog hll = (HyperLogLog) gds;
		for (int i = 0; i < m; i++) {
			if (getBitSetValue(HLLMatrix[i]) < getBitSetValue(hll.HLLMatrix[i]))
				HLLMatrix[i] = hll.HLLMatrix[i];
		}
		return this;
	}
	@Override
	public GeneralDataStructure join(GeneralDataStructure gds,int w,int i) {
		HyperLogLog hll = (HyperLogLog) gds;		
		for(int j=0;j<w;j++) {
			if (getBitSetValue(HLLMatrix[i]) < getBitSetValue(hll.HLLMatrix[i*w+j]))
				HLLMatrix[i] = hll.HLLMatrix[i*w+j];
		}
		return this;
	}
}

File: OVS/datapath/HyperLogLog.c
#include "HyperLogLog.h"
#include "GeneralUtil.h"

HyperLogLog *newHyperLogLog(int m, int size) {
	HyperLogLog *b = malloc(sizeof(HyperLogLog));
	//printf("6!\n");
	//printf("b is %llu", b);
	//if(!b){printf("OOOOOOOOOOOHHHHNOOOOO!\n");}
	//srand(time(NULL));
	b->HLLSize = size;
	//printf("6.1!\n");
	b->m = m;
	//printf("6.2!\n");
	b->HLL = (bool **)kzalloc(sizeof(bool*) * m, GFP_ATOMIC);
	//printf("7!\n");
	int i = 0;
	for (i = 0; i < m; i++) {
	//	printf("8!\n");
		b->HLL[i] = (bool *)malloc(8 * sizeof(bool));
		int j = 0;
		for (j = 0; j < size; j++) {
	//		printf("9!\n");
			b->HLL[i][j] = false;
		}
	}
	b->maxRegisterValue = (int)(31);
	//b->alpha = getAlpha(m);
	// printf("%s\n", "This is a HyperLogLog");
	return b;
}

void encodeHyperLogLog(const HyperLogLog *b) {
	int r = rand();
	int k = r % b->m;
	//unsigned hash_val = (unsigned) hash(distribution(generator));
	unsigned hash_val = uIntHash(r);
	int leadingZeros = numberOfLeadingZeros(hash_val) + 1; // % hyperLogLog_size + hyperLogLog_size) % hyperLogLog_size;
	leadingZeros = min(leadingZeros, b->maxRegisterValue);
	if (getBitsetValue(b->HLL[k]) < leadingZeros) {
		setBitsetValue(b, k, leadingZeros);
	}
	// printf("%s\n", "This is a HyperLogLog");
}

void encodeHyperLogLogEID(const HyperLogLog *b, int elementID) {
	unsigned hash_val = uIntHash((unsigned)elementID);
	int leadingZeros = numberOfLeadingZeros(hash_val) + 1; // % hyperLogLog_size + hyperLogLog_size) % hyperLogLog_size;
	int k = (hash_val % b->m + b->m) % b->m;
	leadingZeros = min(leadingZeros, b->maxRegisterValue);
	if (getBitsetValue(b->HLL[k]) < leadingZeros) {
		setBitsetValue(b, k, leadingZeros);
	}
	//printf("%s\n", "This is a HyperLogLog");
}

void encodeHyperLogLogSegment(const HyperLogLog *b, int flowID, int *s, int w) {
	int ms = b->m / w;
	int r = rand();
	int j = r % ms;			// (GeneralUtil::intHash(flowID) % ms + ms) % ms;
	int k = (intHash(flowID ^ s[j]) % w + w) % w;
	int i = j * w + k;
	unsigned hash_val = uIntHash(r);
	int leadingZeros = numberOfLeadingZeros(hash_val) + 1;
	leadingZeros = min(leadingZeros, b->maxRegisterValue);
	if (getBitsetValue(b->HLL[i]) < leadingZeros) {
		setBitsetValue(b, i, leadingZeros);
	}
}

void encodeHyperLogLogSegmentEID(const HyperLogLog *b, int flowID, int elementID, int *s, int w) {
	int m = b->m / w;
	int j = (intHash(elementID) % m + m) % m;
	int k = (intHash(flowID ^ s[j]) % w + w) % w;
	int i = j * w + k;
	unsigned hash_val = uIntHash(elementID);						// (unsigned)
	int leadingZeros = numberOfLeadingZeros(hash_val) + 1;
	leadingZeros = min(leadingZeros, b->maxRegisterValue);
	if (getBitsetValue(b->HLL[i]) < leadingZeros) {
		setBitsetValue(b, i, leadingZeros);
	}
}

// double getAlpha(int m)
// {
// 	double a;
// 	if (m == 16) {
// 		a = 0.673;
// 	}
// 	else if (m == 32) {
// 		a = 0.697;
// 	}
// 	else if (m == 64) {
// 		a = 0.709;
// 	}
// 	else {
// 		a = 0.7213 / (1 + 1.079 / m);
// 	}
// 	return a;
// }

int getBitsetValue(bool *b)
{
	int result = 0;
	int i = 0;
	for (i = 0; i < 5; i++) {
		if (b[i]) {
			result = result + (1 << i);
		}
	}
	return result;
}

void setBitsetValue(const HyperLogLog *b, int index, int value)
{
	int i = 0;
	while (value != 0 && i < b->HLLSize) {
		if ((value & 1) != 0) {
			b->HLL[index][i] = 1;
		}
		else {
			b->HLL[index][i] = 0;
		}
		value = value >> 1;
		i++;
	}
}
